### DISCUSSION TITLE
Recomended way to do authentification?

### DISCUSSION THREAD

**Discussion by AlexKotik** [Kitware/trame#445] [Q&A]
I'm sorry, if I'm missing something, but is there a recomended way to do user authentification in Trame-based app? I couldn't find docs about it for some reason.

**Comment by jourdain** [comment]
No worries, I guess there is many path depending on what you are looking for.
A better one would be to handle authentication outside so you aren't allowed to run a trame server process before any prior authorization.

Otherwise, you can easily manage it yourself within trame by creating 2 methods that build a layout. You start with your login UI and then if the login/pass are satisfactory, you just execute the second method to build your full app UI.

**Reply by wanqing0421** [reply]
Are there some examples?

**Reply by jourdain** [reply]
Are you asking from within the trame process rather than before hand?

If so something like that

```python
from trame.app import TrameApp
from trame.ui.html import DivLayout
from trame.widgets import html

class AuthDemo(TrameApp):
   def __init__(self):
        super().__init__()
        self.login_ui()

  def check_auth(self):
       if self.state.secret == "valid-secret":
           self.app_ui()

  def login_ui(self):
      with DivLayout(self.server):
          html.Input(v_model=("secret", ""))
          html.Button("Validate", click=self.check_auth)

  def app_ui(self):
      with DivLayout(self.server):
          html.Div("You are logged in")

if __name__ == "__main__":
     app = AuthDemo()
     app.server.start()
    
```
